services/src/modules/directives/gql/introspection.ts

```diff
@@ -15,39 +15,50 @@ export interface RemoteSchema {
   schema: string;
 }
 
-export async function updateRemoteGqlSchemas(
-  schema: GraphQLSchema,
-  resourceGroup: ResourceGroup,
-  activeDirectoryAuth: ActiveDirectoryAuth
-) {
-  const newSchemas = await fetchRemoteGqlSchemas(schema, resourceGroup, activeDirectoryAuth);
-  if (!newSchemas || newSchemas.length === 0) return;
+export async function updateRemoteGqlSchemas(resourceGroup: ResourceGroup, activeDirectoryAuth: ActiveDirectoryAuth) {
+  const knownRemoteGqlServers = new Set(resourceGroup.remoteSchemas?.map(rs => rs.url));
+  const urls = new Set<string>();
+  for (const { schema } of resourceGroup.schemas) {
+    const typeDefs = parse(schema);
+    const objectTypeDefinitions = typeDefs.definitions.filter(
+      d => d.kind === 'ObjectTypeDefinition'
+    ) as ObjectTypeDefinitionNode[];
 
-  if (!resourceGroup.remoteSchemas) {
-    resourceGroup.remoteSchemas = [];
-  }
-  resourceGroup.remoteSchemas.push(...newSchemas);
-}
+    for (const { fields } of objectTypeDefinitions) {
+      if (!fields) continue;
 
-async function fetchRemoteGqlSchemas(
-  schema: GraphQLSchema,
-  resourceGroup: ResourceGroup,
-  activeDirectoryAuth: ActiveDirectoryAuth
-) {
-  const results: Record<string, SchemaDirectiveVisitor[]> = SchemaDirectiveVisitor.visitSchemaDirectives(schema, {
-    gql: GqlRegistryVisitor,
-  });
-  const knownRemoteGqlServers = new Set(resourceGroup.remoteSchemas?.map(rs => rs.url));
-  const gqlIntrospectUrls = results['gql']
-    .map(i => i.args.url)
-    .filter((url, idx, self) => !!url && !knownRemoteGqlServers.has(url) && self.indexOf(url) === idx) as string[];
-  const schemas = await Promise.all(
-    gqlIntrospectUrls.map(async url => ({
-      url,
-      schema: await fetchRemoteGqlSchema(url, resourceGroup, activeDirectoryAuth),
-    }))
-  );
-  return schemas;
+      for (const field of fields) {
+        const gqlDirectives = field.directives?.filter(d => d.name.value === 'gql');
+        if (!gqlDirectives) continue;
+
+        for (const gqlDirective of gqlDirectives) {
+          if (!gqlDirective.arguments) continue;
+
+          const urlArgument = gqlDirective.arguments!.find(a => a.name.value === 'url');
+          if (!urlArgument) continue;
+          const url = (urlArgument!.value as StringValueNode).value;
+
+          if (!knownRemoteGqlServers.has(url)) {
```

Comment:
Doesn't this mean we never update the existing remote schemas? why not update them?

Reply:
I think we should update the remote schemas from time to time. But I'm not sure that the trigger to this update should be update of other not related resource in resource group.